From 71eaacdc1cc0fafe7598c6889c60c09e7f66856f Mon Sep 17 00:00:00 2001
From: Siyeon Seo <siyeon.seo@lge.com>
Date: Wed, 31 Mar 2021 15:18:51 +0900
Subject: [PATCH] Remove unplugged device in knownPointingDevices list

When the visible value of qquickitem is changed, a crash occurs
while accessing the physically unplugged device information from
the knownPointingDevices list.

Pick-to: 6.1
Change-Id: I7f3190bc47ef068ca3d795216eedd6377fa25120
Upstream-Status: Backport [https://codereview.qt-project.org/c/qt/qtdeclarative/+/341438]
---
 src/quick/util/qquickdeliveryagent.cpp | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/quick/util/qquickdeliveryagent.cpp b/src/quick/util/qquickdeliveryagent.cpp
index 44df53cc8a..4fe3e9db6e 100644
--- a/src/quick/util/qquickdeliveryagent.cpp
+++ b/src/quick/util/qquickdeliveryagent.cpp
@@ -1521,10 +1521,12 @@ void QQuickDeliveryAgentPrivate::onGrabChanged(QObject *grabber, QPointingDevice
 
 void QQuickDeliveryAgentPrivate::ensureDeviceConnected(const QPointingDevice *dev)
 {
+    Q_Q(QQuickDeliveryAgent);
     if (knownPointingDevices.contains(dev))
         return;
     knownPointingDevices.append(dev);
     connect(dev, &QPointingDevice::grabChanged, this, &QQuickDeliveryAgentPrivate::onGrabChanged);
+    QObject::connect(dev, &QObject::destroyed, q, [this, dev] {this->knownPointingDevices.removeAll(dev);});
 }
 
 void QQuickDeliveryAgentPrivate::deliverPointerEvent(QPointerEvent *event)
